sdcm/utils/version_utils.py

```diff
@@ -1,68 +1,150 @@
+
+
+class ScyllaFileType(Enum):
+    DEBIAN = auto()
+    YUM = auto()
+
+
+FILE_REGEX_DICT = {
+    ScyllaFileType.DEBIAN: [
+        (re.compile(r"deb\s+\[arch=(?P<arch>.*?)\]\s(?P<url>http.*?)\s(?P<version_code_name>.*?)\s(?P<component>.*?)$"),
+         "{url}/dists/{version_code_name}/{component}/binary-{arch}/Packages"),
+    ],
+    ScyllaFileType.YUM:
+        [
+            (re.compile(r"baseurl=(?P<url>http.*scylladb.com.*)"), "{url}" + REPOMD_XML_PATH),
+            (re.compile(r"deb\s+(?P<url>http.*?)\s(?P<version_code_name>.*?)\s(?P<component>.*?)$"),
+             "{url}/dists/{version_code_name}/{component}/binary-amd64/Packages.gz"),
+    ],
+}
+
+# The variable "type" indices the type of URL (Debian or Yum)
+# The variable "urls" contains all urls using the Scylla pattern
+RepositoryDetails = namedtuple("RepositoryDetails", ["type", "urls"])
+
+
+def get_url_content(url, return_url_data=True):
+    response = requests.get(url=url)
+    if response.status_code != 200:
+        raise ValueError(f"The following repository URL '{url}' is incorrect")
+    response_data = response.text
+    if not response_data:
+        raise ValueError(f"The repository URL '{url}' not contains any content")
+    if return_url_data:
+        return response_data.split('\n')
+    # To overcome on Pylint's "inconsistent-return-statements", a value must be returned    return []
+    return []
+
+
+def get_scylla_urls_from_repository(repo_details):
+    urls = set()
+    for url in repo_details.urls:
+        for (url_regex, url_format) in FILE_REGEX_DICT[repo_details.type]:
+            match = url_regex.match(url)
+            if not match:
+                # Continue until find the correct Regex
+                continue
+
+            full_url = url_format.format(**match.groupdict())
+            # for scylla-manager we never used a noarch key
+            basearch_list = ["x86_64"] if 'scylla-manager' in full_url else ["x86_64", "noarch"]
+            for basearch in basearch_list:
+                urls.add(Template(full_url).substitute(basearch=basearch, releasever='7'))
+            # We found the correct regex and we can continue to next URL
+            break
+
+    urls = list(urls)
+    ParallelObject(objects=urls, timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(func=lambda _url: get_url_content(
+        url=_url, return_url_data=False))
+    return urls
+
+
+def get_branch_version_from_debian_repository(urls):
+    def get_version(url):
+        data = '\n'.join(get_url_content(url=url))
+        # Get only the major version (i.e. "2019.1.1-0.20190709.9f724fedb-1~stretch", get only "2019.1.1")
+        major_versions = [version.split('-', maxsplit=1)[0] for version in REPO_VERSIONS_REGEX.findall(data)]
+        if not major_versions:
+            return ""
+        return max(set(major_versions), key=major_versions.count)
+
+    threads = ParallelObject(objects=urls, timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(func=get_version)
```

Comment:
please for the get version, don't use ParallelObject, it might get us the wrong version, since we are not picking it by name.
think if we have 
```
baseurl=https://repositories.scylladb.com/scylla/downloads/scylladb/qa-test/scylla-enterprise/rpm/centos/scylladb-2019.1/$releasever/$basearch/
baseurl=https://repositories.scylladb.com/scylla/downloads/scylladb/qa-test/scylla-enterprise/rpm/centos/scylladb-2019.1/$releasever/noarch/
baseurl=https://copr-be.cloud.fedoraproject.org/results/scylladb/scylla-3rdparty/epel-7-$basearch/
```
if in each URL we'll have packages with versions like `5.7.9` or `3.8`, and scylla is `3.4`, we might endup with getting `5.7.6` since it the max version

Reply:
No, the function only finds the Scylla versions.
The last line `baseurl=https://copr-be.cloud.fedoraproject.org/results/scylladb/scylla-3rdparty/epel-7-$basearch/` cannot be matched

Reply:
o.k. we'll do with it, the code before that was picking only one url for getting the versions